utils/data.cpp

```diff
@@ -0,0 +1,246 @@
+#include "data.hpp"
+
+namespace jbrush {
+
+Data::Data(Canvas* theCanvas)
+{
+    mainCanvas = theCanvas;
+    currentDirectory = STARTING_DIRECTORY;
+}
+
+void Data::Load()
+{
+    try {
+        // Variables used to custom-craft descriptive error messages in exception handling
+        QString errorMessage;
+        QTextStream errorStream(&errorMessage);
+
+        // Bring up a panel for the user to select a directory
+        SelectDirectory();
+
+        // If directory is not valid, throw an exception
+        if(currentDirectory == "")
+        {
+            errorStream << "Directory" << currentDirectory << "is not valid";
+            GeneralException(errorMessage).raise();
+        }
+
+        // Get the file at the current directory
+        // and make a text stream to operate on it
+        QFile file(currentDirectory);
+        QTextStream qin(&file);
+
+        if(!file.open(QIODevice::ReadOnly))
+        {
+           errorStream << "File at" << currentDirectory << "was not opened properly";
+           GeneralException(errorMessage).raise();
+        }
+
+        QString type;   // Name identifying the shape read in
+        QList<qint32> dimensions;   // This list is used to load in the dimensions of the shape
+        Shape* toAdd;   // The shape to be constructed and added to the canvas
+
+        qint32 tempInt; // Temporary storage for an integer
+        QChar tempChar; // Temporary storage for a character
+
+        // Set integer base. Speeds things up a little for the text stream
+        qin.setIntegerBase(10);
+
+        // Loop and extract data until we reach the end of the file stream
+        while(!qin.atEnd())
+        {
+            GetToWhitespace(qin);
+            qin >> type;
+
+            // Get to next whitespace after leading title
+            GetToWhitespace(qin);
+
+            // Continue reading in integers until one is found that is not followed by the designated dimension interruptor
+            do
+            {
+                qin >> tempInt;
+                dimensions.append(tempInt);
+                qin >> tempChar;
+            }while(tempChar == DIMENSION_INTERRUPTOR);
+
+            // If the shape being read in isn't a text, we know it needs a pen
+            if(type != TEXT)
+            {
+                // Extract the pen
+
+                // If the shape being read in isn't a line, we know it needs a brush
+                if(type != LINE && type != POLYLINE)
+                {
+                    // Extract brush
+
+                    // Try to construct a 2-D.  If successful, add it to the canvas
+                    try {
+                        toAdd = ConstructShape(type, dimensions);
+                        mainCanvas->addShape(toAdd);
```

Comment:
Im not sure if you tested this yet, but in my experience, you had to call ``mainCanvas->repaint();`` if you want to redraw after adding a shape. Actually i recommend calling this after adding all the shapes.